filters/auth/oidc.go

```diff
@@ -860,6 +917,137 @@ func (f *tokenOidcFilter) getTokenWithExchange(state *OauthState, ctx filters.Fi
+// 	   }
+//   }
+// }
+//
+func (f *tokenOidcFilter) handleDistributedClaims(ctx context.Context, idToken *oidc.IDToken, oauth2Token *oauth2.Token, claimsMap map[string]interface{}) error {
+	// https://github.com/coreos/go-oidc/issues/171#issuecomment-1044286153
+	var distClaims distributedClaims
+	err := idToken.Claims(&distClaims)
+	if err != nil {
+		return err
+	}
+	if len(distClaims.ClaimNames) == 0 || len(distClaims.ClaimSources) == 0 {
+		log.Debugf("No distributed claims found")
+		return nil
+	}
+
+	for claim, ref := range distClaims.ClaimNames {
+		source, ok := distClaims.ClaimSources[ref]
+		if !ok {
+			return fmt.Errorf("invalid distributed claims: missing claim source for %s", claim)
+		}
+		uri, err := url.Parse(source.Endpoint)
+		if err != nil {
+			return fmt.Errorf("failed to parse distributed claim endpoint: %w", err)
+		}
+
+		var results []interface{}
+
+		switch uri.Host {
+		case azureADGraphHost, microsoftGraphHost:
+			results, err = f.handleDistributedClaimsAzure(uri, oauth2Token, claimsMap)
+			if err != nil {
+				return fmt.Errorf("failed to get distributed Azure claim: %w", err)
+			}
+		default:
+			return fmt.Errorf("unsupported distributed claims endpoint '%s', please create an issue at https://github.com/zalando/skipper/issues/new/choose", uri.Host)
+		}
+
+		claimsMap[claim] = results
+	}
+	return nil
+}
+
+// Azure customizations https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-overview
+// If the endpoints provided in _claim_source is pointed to the deprecated "graph.windows.net" api
+// replace with handcrafted url to graph.microsoft.com
+func (f *tokenOidcFilter) handleDistributedClaimsAzure(url *url.URL, oauth2Token *oauth2.Token, claimsMap map[string]interface{}) (values []interface{}, err error) {
+	url.Host = microsoftGraphHost
+	// transitiveMemberOf for group names
+	userID, ok := claimsMap["oid"].(string)
+	if !ok {
+		return nil, fmt.Errorf("oid claim not found in claims map")
+	}
+	url.Path = fmt.Sprintf("/v1.0/users/%s/transitiveMemberOf", userID)
+	q := url.Query()
+	q.Set("$select", "displayName,id")
+	url.RawQuery = q.Encode()
+	return f.resolveDistributedClaimAzure(url, oauth2Token)
+}
+
+func (f *tokenOidcFilter) resolveDistributedClaimAzure(url *url.URL, oauth2Token *oauth2.Token) (values []interface{}, err error) {
+	var target azureGraphGroups
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error constructing groups endpoint request: %w", err)
+	}
+	oauth2Token.SetAuthHeader(req)
+
+	cli, _ := distributedClaimsClients.LoadOrStore(url.Host, snet.NewClient(snet.Options{
+		ResponseHeaderTimeout:   f.oidcOptions.Timeout,
+		TLSHandshakeTimeout:     f.oidcOptions.Timeout,
+		MaxIdleConnsPerHost:     f.oidcOptions.MaxIdleConns,
+		Tracer:                  f.oidcOptions.Tracer,
+		OpentracingComponentTag: "skipper",
+		OpentracingSpanName:     "distributedClaims",
+	}))
```

Comment:
@AlexanderYastrebov yes true snet.NewClient should not be called everytime. 
What's your reasoning to store client per url?
disclaimer: I don't think it makes sense per url, because TCP connection pools will resolve DNS and connect to the resolved IPs, whatever url you have. So it makes sense to have per host, because of this mapping. If people will hit a limit of connections that creates latency they can increase MaxIdleConnsPerHost and call it a day.

Reply:
Don't think storing on URL key would be a good idea as URLs are usually one-time use, which essentially results into a  leak.

Reply:
I split the LoadAndRestore into conditioned sections to prevent creation of a client instance.

I'd prefer to keep using the `sync.Map` here as it is faster and well tested:

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

Reply:
> What's your reasoning to store client per url?

I think we do per URL elsewhere e.g. for tokeninfo clients. 
Lets store per host given your pros arguments.

> I split the LoadAndRestore into conditioned sections to prevent creation of a client instance.

Which introduces a race condition. When there is no mapping stored it could happen that two request create a client each, one will be stored and another will leak. Presence check and addition should be done while the map is locked for updates - that is `LoadOrStore` for (unfortunately it lacks capability of lazy value initialization https://github.com/golang/go/issues/44159)

Reply:
To avoid leaking the client I think we would need something like:
```go
	cli, ok := distributedClaimsClients.Load(url.Host)
	if !ok {
		newCli = initClient()
		cli, loaded = distributedClaimsClients.LoadOrStore(url.Host, newCli)
		if loaded {
			newCli.Close()
		}
	}
```

Reply:
> To avoid leaking the client I think we would need something like:

True, that is the proper way, thanks.

> I think we do per URL elsewhere e.g. for tokeninfo clients.
Lets store per host given your pros arguments.

Thinking twice, why do we store the client as instances (per host) at the first place, as we do not have any host part stored with it? Logically, every client instance is going to be equal, wouldn't a single static client be just fine? @AlexanderYastrebov @szuecs

Reply:
In case of oidc it should be fine anyways. I am not sure anymore without reading the code why we have a client per url/host in case of tokeninfo/tokenintrospection , I think it's because we have low latency conditions (or it's about reloading configuration) and oidc is pretty much slow anyways (2-4 requests with redirects for one flow), so it doesn't really matter.

Reply:
sorry the delay!
Thanks for the context @szuecs - then it should not hurt to keep it like it is.

@AlexanderYastrebov implemented your suggestion, please check